File: Data Structures & Algorithms/subsets-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    // time - O(n * 2^n), space - O(n * 2^n)
    subsetsWithDup(nums) {
        const subsets = []
        nums.sort((a,b) => a-b)
        const recursion = (index, subset) => {
            subsets.push([...subset])
            for (let i = index; i < nums.length; i++) {
                if (i > index && nums[i] === nums[i-1]) continue
                subset.push(nums[i])
                recursion(i+1, subset)
                subset.pop()
            }
        }
        recursion(0, [])
        return subsets
    }
}

File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    private List<Integer> subset = new ArrayList<>();
    private List<List<Integer>> output = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtrack(nums, 0);
        return output;
    }

    private void backtrack(int[] nums, int index) {
        if (index == nums.length) {
            output.add(new ArrayList<>(subset));
            return;
        }

        subset.add(nums[index]);
        backtrack(nums, index + 1);
        subset.remove(subset.size() - 1);
        backtrack(nums, index + 1);
    }
}